File: Bi_directional/Graph.py
import math
class Graph:
	def __init__(self, point):
		self.root = Node(point)
		self.lastNode = None

	def findNearest(self, node, point): #find the nearest node form the given point
		arr = [node, node.point.Euc_Dist(point)]
		for child in node.children:
			arr_temp = self.findNearest(child, point) #recursion
			if arr_temp[1] < arr[1]:
				arr = arr_temp
		return arr

	def addNode(self, point):
		node = self.findNearest(self.root, point)[0]
		newNode = Node(point)
		node.addNode(newNode)
		return newNode

	def add_lastNode(self, node):
		self.lastNode = node


	def traverseGraph(self, node): #traversal system(for my understanding)
		node.point.disp()
		print("Now children")
		for child in node.children:
			self.traverseGraph(child)

		print("END children")

	def findPath(self):
		points_Arr = []
		node = self.lastNode
		while(node!=None):
			points_Arr.append([node.point.x, node.point.y])
			node = node.parent
		return points_Arr





class Node:
	def __init__(self, point):
		self.point = point
		self.children = []
		self.parent = None

	def addNode(self, node):
		self.children.append(node)
		node.parent = self

File: RRT/Graph.cpp
#include "Graph.h"
#include <iostream>

using namespace std;

Graph::Graph(Node* node) {
	this->rootNode = node;
}

Node* Graph::findNearest(point2D* point) {
	auto [closestNode, dist] = findNearest_internal(rootNode, point);
	return closestNode;
}

pair<Node*, double>  Graph::findNearest_internal(Node* node, point2D* point) {
	double cur_minDist = node->point->Euc_Dist(point);
	Node* closestNode = node;
	for (int i = 0; i < node->children.size(); i++) {
		auto [childNode, child_dist] = findNearest_internal(node->children[i], point);
		if (child_dist < cur_minDist) {
			closestNode = childNode;
			cur_minDist = child_dist;
		}
	}
	return {closestNode, cur_minDist};
}

Node* Graph::addNode(point2D* point) {
	Node* closestNode = findNearest(point);
	Node* newNode = new Node(point);
	closestNode->connectNode(newNode);
	return newNode;
}

void Graph::setLastNode(Node* node) {
	lastNode = node;
}

void Graph::traversal(Node* node) {
	node->point->disp();
	cout << "Now children" << endl;
	for (int i = 0; i < node->children.size(); i++) {
		traversal(node->children[i]);
	}
	cout << "END children"<< endl;
}

pair<vector<double>, vector<double>> Graph::retracePath() {
	vector<double> path_x;
	vector<double> path_y;
	Node* node = this->lastNode;
	while(node != NULL) {
		// cout << "NOW AT " << flush;
		// node->point->disp();
		path_x.push_back(node->point->x);
		path_y.push_back(node->point->y);
		node = node->parent;
	}
	return {path_x, path_y};
}

// Graph::~Graph() {
// 	delete rootNode;
// 	delete lastNode;
// }